### ISSUE TITLE
`argparse` does not correctly handle a default set to `SUPPRESS` when using `set_defaults()`

### ISSUE THREAD

**Issue by daggaz** [python/cpython#107620]
# Bug report

## Checklist

- [X] I am confident this is a bug in CPython, not a bug in a third-party project
- [X] I have searched the CPython issue tracker, and am confident this bug has not been reported before

## A clear and concise description of the bug

`argparse` does not correctly handle a default set to `SUPPRESS` when the default is supplied via `set_defaults()`, and instead outputs `'==SUPPRESS==`` into the resulting `Namespace`.

I would expect the following three snippets to work identically:
```python
import argparse

# set default=SUPPRESS via add_argument()
parser = argparse.ArgumentParser()
print(parser.add_argument('--foo', default=argparse.SUPPRESS))
print(parser.parse_args([]))

# set default=SUPPRESS via argument_default
print()
parser = argparse.ArgumentParser(argument_default=argparse.SUPPRESS)
print(parser.add_argument('--foo'))
print(parser.parse_args([]))

# set default=SUPPRESS via set_defaults
print()
parser = argparse.ArgumentParser()
parser.set_defaults(foo=argparse.SUPPRESS)
print(parser.add_argument('--foo'))
print(parser.parse_args([]))
```

In the last case, the namespace comes out as:

```python
Namespace(foo='==SUPPRESS==')
```

The issue appears to be in `ArgumentParser.parse_known_args()` when it processes the `_defaults` dictionary.

In this code ([Lib/argparse.py#L1919](https://github.com/python/cpython/blob/2ba7c7f7b151ff56cf12bf3cab286981bb646c90/Lib/argparse.py#L1919)):
```python
# add any action defaults that aren't present
for action in self._actions:
    if action.dest is not SUPPRESS:
        if not hasattr(namespace, action.dest):
            if action.default is not SUPPRESS:
                setattr(namespace, action.dest, action.default)

# add any parser defaults that aren't present
for dest in self._defaults:
    if not hasattr(namespace, dest):
        setattr(namespace, dest, self._defaults[dest])
```

The first loop processes all the registered actions, checking for those missing from the namespace, and skipping any suppressed ones. In the second loop, the suppression logic is missing.

# Your environment

Python 3.11.4, MacBook Pro, 14-inch, 2021, Apple M1 Pro

### Linked PRs
* gh-107621

**Comment by hpaulj** [comment]
I don't think this is a bug.  It's had this behavior since the begining, and it's intentional.  The documentation for `set_defaults` says

[set_defaults()](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.set_defaults) allows some additional attributes that are determined without any inspection of the command line to be added:

The main documented use is to add a `function` attribute to subparsers.

While it can be used override the default of a defined argument, mostly it is used to add an attribute to the namespace that is not linked to any `add_argument` call.  As such using SUPPRESS doesn't make much sense.  Why add such default only to suppress it?  If you don't want such an attribute in the namespace, don't use `set_defaults`.

If you previuosly added an argument, e.g.

     a1 = parser.add.argument(...)

and want to change its default, you can just do

    a1.default = argparse.SUPPRESS

(or if you didn't save the `a1` reference you can find it in the `._actions` list.)

I have seen people try to change the defaults of Actions added via parents.  But that doesn't work well, because Actions are 'copied' by reference.

**Comment by daggaz** [comment]
In my particular use, I'm building a set of defaults whose values change depending on certain conditions (in fact an earlier pass of `argparse` with a reduced set of arguments and `parse_known_args()`).

Building this as a dictionary to pass to `set_defaults()` seems natural to me, and the fact the `SUPPRESS` was handled differently surprised me.

I agree I _could_, for example, iterate my dictionary, and edit the action objects directly, but I raised this case because I was surprised that this didn't just work.

As an aside, my use case, is not about the suppression of the value in the resulting namespace. I was using `SUPPRESS` in conjunction with `ArgumentDefaultsHelpFormatter` to hide the "(default: ...)" help text when my dictionary did not supply a value.

**Comment by hpaulj** [comment]
Another nuance.  If we use `ArgumentDefaultHelpFormatter` to add the `default` display, the `SUPPRESS` blocks that addition.

But if we explicitly include the default display in the help, such as with:

     p._actions[1].help='foo %(default)s'

`p.print_help()` raises a key-error.  That's because the formatter `_expand_help` method pays attention to the `SUPPRESS`.  In ordinary usage that's ok, because you wouldn't `SUPPRESS` the default and also ask the help to show the default.  

Anyways, 'automating' `argparse` setup with can be tricky because it can create a combination of parameters that the original developers did not plan for.

I wonder if your case, a custom version of `set_defaults` would be useful, one which sets the `Action` defaults, but does not update the `._defaults` dict.  That could be an independent function that just iterates on `parser._actions`.  

`set_defaults` can serve two purposes.  1) easily setting or changing the default of an Action based on its `dest`, 2) creating a namespace default that's independent of any Action.  You have enountered a case in which those two purposes create a conflict.
